sw/airborne/firmwares/rotorcraft/guidance/guidance_indi_hybrid.c

```diff
@@ -309,8 +328,41 @@ struct StabilizationSetpoint guidance_indi_run(struct FloatVect3 *accel_sp, floa
 #endif
 #endif
 
-  //Calculate roll,pitch and thrust command
-  MAT33_VECT3_MUL(euler_cmd, Ga_inv, a_diff);
+  // Calculate matrix of partial derivatives
+  guidance_indi_calcg_wing(Ga);
+
+#if GUIDANCE_INDI_HYBRID_USE_WLS
+  // Set lower limits
+  du_min_gih[0] = -guidance_indi_max_bank - roll_filt.o[0]; // roll
+  du_min_gih[1] = RadOfDeg(guidance_indi_min_pitch) - pitch_filt.o[0]; // pitch
+  du_min_gih[2] = (MAX_PPRZ - actuator_state_filt_vect[0]) * g1g2[3][0] + (MAX_PPRZ - actuator_state_filt_vect[1]) * g1g2[3][1] + (MAX_PPRZ - actuator_state_filt_vect[2]) * g1g2[3][2] + (MAX_PPRZ - actuator_state_filt_vect[3]) * g1g2[3][3];
+
+  // Set upper limits limits
+  du_max_gih[0] = guidance_indi_max_bank - roll_filt.o[0]; //roll
+  du_max_gih[1] = RadOfDeg(GUIDANCE_INDI_MAX_PITCH) - pitch_filt.o[0]; // pitch
+  du_max_gih[2] = -(actuator_state_filt_vect[0]*g1g2[3][0] + actuator_state_filt_vect[1]*g1g2[3][1] + actuator_state_filt_vect[2]*g1g2[3][2] + actuator_state_filt_vect[3]*g1g2[3][3]);
+
+  // Set prefered states
+  du_pref_gih[0] = 0.; // prefered delta roll angle
```

Comment:
I think it would be better to make this ```- roll_filt.o[0]```. By doing this we put some cost on rolling away from 0, which will be helpful if you have direct lateral thrust control. In that case and with the desired delta set to zero, you could drift to a case where you are rolling and compensating the roll with lateral thrust, with no encouragement to be energy efficient.

Reply:
Solved in https://github.com/paparazzi/paparazzi/tree/rebase_wls